packages/env/lib/config/post-process-config.js

```diff
@@ -0,0 +1,195 @@
+	// cascade behavior that would break the normal merge. After merging we then
+	// delete them to avoid that breakage and add them back before we return.
+	const removedRootOptions = {};
+	if (
+		config.port !== undefined &&
+		config.env.development.port === undefined
+	) {
+		removedRootOptions.port = config.port;
+		config.env.development.port = config.port;
+		delete config.port;
+	}
+	if (
+		config.testsPort !== undefined &&
+		config.env.tests.port === undefined
+	) {
+		removedRootOptions.testsPort = config.testsPort;
+		config.env.tests.port = config.testsPort;
+		delete config.testsPort;
+	}
+
+	// Merge the root config and the environment configs together so that
+	// we can ignore the root config and have full environment configs.
+	for ( const env in config.env ) {
+		config.env[ env ] = mergeConfigs(
+			deepCopyRootOptions( config ),
+			config.env[ env ]
+		);
+	}
+
+	// Set any root-level options we reset back.
+	for ( const option in removedRootOptions ) {
+		config[ option ] = removedRootOptions[ option ];
+	}
+
+	return config;
+}
+
+/**
+ * Appends the configured port to certain wp-config options.
+ *
+ * @param {WPServiceConfig} config The config to process.
+ *
+ * @return {WPServiceConfig} The config after post-processing.
+ */
+function appendPortToWPConfigs( config ) {
+	const options = [ 'WP_TESTS_DOMAIN', 'WP_SITEURL', 'WP_HOME' ];
+
+	// We are only interested in editing the config options for environment-specific configs.
+	// If we made this change to the root config it would cause problems since they would
+	// be mapped to all environments even though the ports will be different.
+	for ( const env in config.env ) {
+		// There's nothing to do without any wp-config options set.
+		if ( config.env[ env ].config === undefined ) {
+			continue;
+		}
+
+		if ( config.env[ env ].port === undefined ) {
+			continue;
+		}
+
+		// Make sure that the port is on the option if it's present.
+		for ( const option of options ) {
+			if ( config.env[ env ].config[ option ] === undefined ) {
+				continue;
+			}
+
+			config.env[ env ].config[ option ] = addOrReplacePort(
+				config.env[ env ].config[ option ],
+				config.env[ env ].port,
+				// Don't replace the port if one is already set on WP_HOME.
+				option !== 'WP_HOME'
+			);
+		}
+	}
+
+	return config;
+}
+
+/**
+ * Examines the config to make sure that none of the environments share the same port.
+ *
+ * @param {WPServiceConfig} config The config to process.
+ */
+function validatePortUniqueness( config ) {
+	// We're going to build a map of the environments and their port
+	// so we can accomodate root-level config options more easily.
+	const environmentPorts = {};
+
+	// Add all of the environments to the map. This will
+	// overwrite any root-level options if necessary.
+	for ( const env in config.env ) {
+		if ( config.env[ env ].port === undefined ) {
+			throw new ValidationError( `The "${ env }" has an invalid port.` );
```

Comment:
I think this would come out as "The development has an invalid port" -- might need some clarification :) 

side note -- is it possible to get into this condition at this point?

Reply:
You're right, it would. As for the condition itself, no, I don't think so. It was possible in an earlier version of this class, but, I don't think so now. I'd rather leave the extra validation though, just in case :)